File: systems-experiments/erdos-experiments/src/bin/synthetic_pipeline.rs
use clap::Arg;
use erdos::{
    dataflow::{
        operators::{Concat, FlatMapOperator},
        stream::{ExtractStream, Stream},
    },
    node::{Node, NodeId},
    Configuration, OperatorConfig,
};
use erdos_experiments::{
    messages::{BinaryData, LatencySample},
    operators::{binary_source::BinarySource, csv_sink::CSVSink, join::JoinOperator},
};

// Typical sizes of messages, taken from a trace of a Pylot run.
const CAMERA_MSG_SIZE: usize = 921600; // 6222274;
const LIDAR_MSG_SIZE: usize = 600_000;
const LOCALIZATION_MSG_SIZE: usize = 1_000;
const DETECTION_MSG_SIZE: usize = 4433;
const TRAFFIC_LIGHT_MSG_SIZE: usize = 246;
const LANE_DETECTION_MSG_SIZE: usize = 358;
const TRACKER_MSG_SIZE: usize = 123;
const PREDICTION_MSG_SIZE: usize = 600;
const PLANNING_MSG_SIZE: usize = 13099;
const CONTROL_MSG_SIZE: usize = 140;
const FUSION_MSG_SIZE: usize = 10_000;

fn connect_map(
    name: &str,
    msg_size: usize,
    assigned_node: NodeId,
    is_sensor: bool,
    stream: &Stream<BinaryData>,
) -> Stream<BinaryData> {
    erdos::connect_one_in_one_out(
        move || {
            FlatMapOperator::new(move |input_data: &BinaryData| {
                let mut output_data = BinaryData::new(msg_size);
                if !is_sensor {
                    output_data.send_time = input_data.send_time
                }
                std::iter::once(output_data)
            })
        },
        || {},
        OperatorConfig::new().name(name).node(assigned_node),
        stream,
    )
}

fn connect_join(
    name: &str,
    msg_size: usize,
    assigned_node: NodeId,
    left_stream: &Stream<BinaryData>,
    right_stream: &Stream<BinaryData>,
) -> Stream<BinaryData> {
    erdos::connect_two_in_one_out(
        move || {
            JoinOperator::new(move |x: &[BinaryData], y: &[BinaryData]| {
                if x.len() == 0 || y.len() == 0 {
                    return vec![];
                }
                assert_eq!(x.len(), 1);
                assert_eq!(y.len(), 1);
                let mut data = BinaryData::new(msg_size);
                data.send_time = x[0].send_time.min(y[0].send_time);
                vec![data]
            })
        },
        || {},
        OperatorConfig::new().node(assigned_node).name(name),
        left_stream,
        right_stream,
    )
}

fn setup_dataflow(
    pipeline_copies: usize,
    send_frequency: f32,
    samples: usize,
    warmup_samples: usize,
    csv_filename: String,
    node_id: NodeId,
    is_inter_process: bool,
) -> Option<ExtractStream<()>> {
    let node_increment = if is_inter_process { 1 } else { 0 };

    let mut assigned_node: NodeId = 0;
    let mut sample_streams = vec![];

    // Synchronously signals other operators to produce data.
    let source_stream = erdos::connect_source(
        move || BinarySource::new(0, warmup_samples, samples, send_frequency),
        || {},
        OperatorConfig::new().node(assigned_node),
    );
    for _ in 0..pipeline_copies {
        let camera_1_stream = connect_map(
            "Camera 1",
            CAMERA_MSG_SIZE,
            assigned_node,
            true,
            &source_stream,
        );
        assigned_node += node_increment;

        let camera_2_stream = connect_map(
            "Camera 2",
            CAMERA_MSG_SIZE,
            assigned_node,
            true,
            &source_stream,
        );
        assigned_node += node_increment;

        let lidar_stream =
            connect_map("LIDAR", LIDAR_MSG_SIZE, assigned_node, true, &source_stream);
        assigned_node += node_increment;

        let localization_stream = connect_map(
            "Localization",
            LOCALIZATION_MSG_SIZE,
            assigned_node,
            true,
            &source_stream,
        );
        assigned_node += node_increment;

        let _segmentation_stream = connect_map(
            "Segmentation",
            CAMERA_MSG_SIZE,
            assigned_node,
            false,
            &camera_1_stream,
        );
        assigned_node += node_increment;

        let detection_stream = connect_map(
            "Obstacle detection",
            DETECTION_MSG_SIZE,
            assigned_node,
            false,
            &camera_1_stream,
        );
        assigned_node += node_increment;

        let traffic_light_detection_stream = connect_map(
            "Traffic light detection",
            TRAFFIC_LIGHT_MSG_SIZE,
            assigned_node,
            false,
            &camera_2_stream,
        );
        assigned_node += node_increment;

        let lane_detection_stream = connect_map(
            "Lane detection",
            LANE_DETECTION_MSG_SIZE,
            assigned_node,
            false,
            &camera_2_stream,
        );
        assigned_node += node_increment;

        let tracker_stream = connect_join(
            "Tracker",
            TRACKER_MSG_SIZE,
            assigned_node,
            &camera_1_stream,
            &detection_stream,
        );
        assigned_node += node_increment;

        let lidar_traffic_light_fusion_stream = connect_join(
            "LIDAR traffic light fusion",
            FUSION_MSG_SIZE,
            assigned_node,
            &lidar_stream,
            &traffic_light_detection_stream,
        );
        assigned_node += node_increment;

        let tmp_fusion_stream = connect_join(
            "tmp fusion",
            FUSION_MSG_SIZE,
            assigned_node,
            &tracker_stream,
            &lidar_stream,
        );
        let main_fusion_stream = connect_join(
            "Fusion",
            FUSION_MSG_SIZE,
            assigned_node,
            &tmp_fusion_stream,
            &localization_stream,
        );
        assigned_node += node_increment;

        let prediction_stream = connect_join(
            "Prediction",
            PREDICTION_MSG_SIZE,
            assigned_node,
            &main_fusion_stream,
            &localization_stream,
        );
        assigned_node += node_increment;

        let tmp_planning_stream = connect_join(
            "tmp planning",
            PLANNING_MSG_SIZE,
            assigned_node,
            &prediction_stream,
            &lane_detection_stream,
        );
        let planning_stream = connect_join(
            "Planning",
            PLANNING_MSG_SIZE,
            assigned_node,
            &tmp_planning_stream,
            &lidar_traffic_light_fusion_stream,
        );
        assigned_node += node_increment;

        let control_stream = connect_join(
            "Control",
            CONTROL_MSG_SIZE,
            assigned_node,
            &planning_stream,
            &localization_stream,
        );

        assigned_node += node_increment;
        let sample_stream = erdos::connect_one_in_one_out(
            move || {
                FlatMapOperator::new(move |x: &BinaryData| {
                    let recv_time = erdos_experiments::ns_since_unix_epoch();
                    let latency_sample = LatencySample {
                        latency_secs: (recv_time - x.send_time) as f64 / 1e9,
                        send_time: x.send_time as f64 / 1e9,
                        recv_time: recv_time as f64 / 1e9,
                        msg_size: x.size(),
                        is_inter_process,
                    };

                    std::iter::once(latency_sample)
                })
            },
            || {},
            OperatorConfig::new().node(assigned_node),
            &control_stream,
        );
        sample_streams.push(sample_stream);
        assigned_node += node_increment;
    }

    // Merging streams and writing to CSV can occur on the same node
    // (defaults to node 0).
    let init = sample_streams.pop().unwrap();
    let all_samples = sample_streams.into_iter().fold(init, |x, y| x.concat(&y));

    // operators::connect_csv_sink(assigned_node, warmup_samples, csv_filename, sample_stream);
    let done_stream = erdos::connect_one_in_one_out(
        move || CSVSink::new(&csv_filename),
        || {},
        OperatorConfig::new().node(assigned_node),
        &all_samples,
    );

    if node_id == 0 {
        Some(ExtractStream::new(&done_stream))
    } else {
        None
    }
}

fn main() {
    let args = erdos::new_app("broadcast-latency")
        .arg(
            Arg::with_name("frequency")
                .short("f")
                .long("frequency")
                .takes_value(true)
                .default_value("10"),
        )
        .arg(
            Arg::with_name("samples")
                .short("n")
                .long("samples")
                .takes_value(true)
                .default_value("10000"),
        )
        .arg(
            Arg::with_name("warmup-samples")
                .short("w")
                .long("warmup-samples")
                .takes_value(true)
                .default_value("100"),
        )
        .arg(
            Arg::with_name("pipeline-copies")
                .short("p")
                .long("pipeline-copies")
                .takes_value(true)
                .default_value("1"),
        )
        .arg(
            Arg::with_name("output")
                .short("o")
                .long("output")
                .help("Output CSV filename")
                .takes_value(true)
                .required(true),
        )
        .get_matches();

    let send_frequency: f32 = args.value_of("frequency").unwrap().parse().unwrap();
    let samples: usize = args.value_of("samples").unwrap().parse().unwrap();
    let warmup_samples: usize = args.value_of("warmup-samples").unwrap().parse().unwrap();
    let pipeline_copies: usize = args.value_of("pipeline-copies").unwrap().parse().unwrap();
    let output: String = args.value_of("output").unwrap().parse().unwrap();
    let node_id: NodeId = args.value_of("index").unwrap().parse().unwrap();
    let data_addresses: Vec<_> = args
        .value_of("data-addresses")
        .unwrap()
        .split(",")
        .collect();
    let is_inter_process = data_addresses.len() > 1;

    if is_inter_process {
        assert_eq!(
            data_addresses.len(),
            15 * pipeline_copies + 1,
            "Must launch 15 * pipeline_copies + 1 nodes to run in inter-process mode!",
        );
    }

    let mut node = Node::new(Configuration::from_args(&args));

    let extract_stream_opt = setup_dataflow(
        pipeline_copies,
        send_frequency,
        samples,
        warmup_samples,
        output,
        node_id,
        is_inter_process,
    );

    if let Some(mut extract_stream) = extract_stream_opt {
        let node_handle = node.run_async();

        loop {
            let msg = extract_stream.read().unwrap();
            if msg.is_top_watermark() {
                tracing::info!("Shutting down ERDOS.");
                break;
            }
        }

        node_handle.shutdown().unwrap();
    } else {
        node.run();
    }
}

File: systems-experiments/flink-experiments/src/main/java/flink_experiments/SyntheticPipeline.java
package flink_experiments;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Vector;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction;
import org.apache.flink.util.Collector;

import flink_experiments.messages.*;
import flink_experiments.operators.*;

public class SyntheticPipeline {
    final static int CAMERA_MSG_SIZE = 921600; // 6222274;
    final static int LIDAR_MSG_SIZE = 600_000;
    final static int LOCALIZATION_MSG_SIZE = 1_000;
    final static int DETECTION_MSG_SIZE = 4433;
    final static int TRAFFIC_LIGHT_MSG_SIZE = 246;
    final static int LANE_DETECTION_MSG_SIZE = 358;
    final static int TRACKER_MSG_SIZE = 123;
    final static int PREDICTION_MSG_SIZE = 600;
    final static int PLANNING_MSG_SIZE = 13099;
    final static int CONTROL_MSG_SIZE = 140;
    final static int FUSION_MSG_SIZE = 10_000;

    public static void main(String[] args) throws Exception {
        Options options = new Options();

        Option pipeline_copies_arg = new Option("c", "pipeline-copies", true, "Number of pipelines to run.");
        pipeline_copies_arg.setRequired(true);
        options.addOption(pipeline_copies_arg);

        Option frequency_arg = new Option("f", "frequency", true, "Frequency");
        frequency_arg.setRequired(true);
        options.addOption(frequency_arg);

        Option samples_arg = new Option("n", "samples", true, "Number of samples to take.");
        samples_arg.setRequired(true);
        options.addOption(samples_arg);

        Option warmup_samples_arg = new Option("w", "warmup-samples", true, "Number of samples used to warm up");
        warmup_samples_arg.setRequired(true);
        options.addOption(warmup_samples_arg);

        Option is_inter_process_arg = new Option("i", "inter-process", false,
                "Signals that messages should be sent across processes");
        options.addOption(is_inter_process_arg);

        Option output_arg = new Option("o", "output", true, "Output CSV filename");
        output_arg.setRequired(true);
        options.addOption(output_arg);

        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();
        CommandLine cmd = null;

        try {
            cmd = parser.parse(options, args);
        } catch (ParseException e) {
            System.out.println(e.getMessage());
            formatter.printHelp("SyntheticPipeline", options);

            System.exit(1);
        }

        System.out.println(cmd.getOptionValue("pipeline-copies"));
        final int pipeline_copies = Integer.parseInt(cmd.getOptionValue("pipeline-copies"));
        final double send_frequency = Double.parseDouble(cmd.getOptionValue("frequency"));
        final int samples = Integer.parseInt(cmd.getOptionValue("samples"));
        final int warmup_samples = Integer.parseInt(cmd.getOptionValue("warmup-samples"));
        final String csv_filename = cmd.getOptionValue("output");
        final boolean is_inter_process = cmd.hasOption("inter-process");

        SyntheticPipeline.run_experiment(pipeline_copies, send_frequency, samples, warmup_samples,
                csv_filename, is_inter_process);
    }

    private static DataStream<BinaryData> connect_map(DataStream<BinaryData> stream, final String name,
            final int msg_size,
            final boolean is_sensor,
            boolean is_inter_process) {
        String slot_sharing_group = is_inter_process ? name : "Source";
        return stream.map(new MapFunction<BinaryData, BinaryData>() {
            @Override
            public BinaryData map(BinaryData value) throws Exception {
                BinaryData data = new BinaryData(msg_size);
                if (!is_sensor) {
                    data.send_time_ns = value.send_time_ns;
                }
                return data;
            }
        }).slotSharingGroup(slot_sharing_group);
    }

    private static DataStream<BinaryData> connect_join(DataStream<BinaryData> left_stream,
            DataStream<BinaryData> right_stream, final String name, final int msg_size, boolean is_inter_process) {
        String slot_sharing_group = is_inter_process ? name : "Source";
        return left_stream.connect(right_stream).flatMap(new CoFlatMapFunction<BinaryData, BinaryData, BinaryData>() {
            Deque<BinaryData> left_msgs = new ArrayDeque<BinaryData>();
            Deque<BinaryData> right_msgs = new ArrayDeque<BinaryData>();

            @Override
            public void flatMap1(BinaryData left_msg, Collector<BinaryData> out) throws Exception {
                BinaryData right_msg = this.right_msgs.pollFirst();
                if (right_msg != null) {
                    BinaryData data = new BinaryData(msg_size);
                    data.send_time_ns = Math.min(left_msg.send_time_ns, right_msg.send_time_ns);
                    out.collect(data);
                } else {
                    this.left_msgs.addLast(left_msg);
                }
            }

            @Override
            public void flatMap2(BinaryData right_msg, Collector<BinaryData> out) throws Exception {
                BinaryData left_msg = this.left_msgs.pollFirst();
                if (left_msg != null) {
                    BinaryData data = new BinaryData(msg_size);
                    data.send_time_ns = Math.min(left_msg.send_time_ns, right_msg.send_time_ns);
                    out.collect(data);
                } else {
                    this.right_msgs.addLast(right_msg);
                }
            }
        }).slotSharingGroup(slot_sharing_group);
    }

    private static void run_experiment(int pipeline_copies, double send_frequency, int num_msgs,
            int num_warmup_msgs, String csv_filename, final boolean is_inter_process) {

        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setBufferTimeout(0L);

        // Single source to synchronize sensors.
        DataStream<BinaryData> source_stream = env
                .addSource(new BinarySource(0, num_warmup_msgs, num_msgs, send_frequency), "Source")
                .slotSharingGroup("Source");

        Vector<DataStream<LatencySample>> latency_streams = new Vector<DataStream<LatencySample>>();

        for (int pipeline_copy = 0; pipeline_copy < pipeline_copies; pipeline_copy++) {

            DataStream<BinaryData> camera_1_stream = connect_map(source_stream,
                    String.format("Pipeline %d: Camera 1", pipeline_copy), CAMERA_MSG_SIZE, true, is_inter_process);

            DataStream<BinaryData> camera_2_stream = connect_map(source_stream,
                    String.format("Pipeline %d: Camera 2", pipeline_copy), CAMERA_MSG_SIZE, true, is_inter_process);

            DataStream<BinaryData> lidar_stream = connect_map(source_stream,
                    String.format("Pipeline %d: LIDAR", pipeline_copy), LIDAR_MSG_SIZE, true, is_inter_process);

            DataStream<BinaryData> localization_stream = connect_map(source_stream,
                    String.format("Pipeline %d: Localization", pipeline_copy), LOCALIZATION_MSG_SIZE, true,
                    is_inter_process);

            DataStream<BinaryData> segmentation_stream = connect_map(camera_1_stream,
                    String.format("Pipeline %d: Segmentation", pipeline_copy), CAMERA_MSG_SIZE, false,
                    is_inter_process);

            DataStream<BinaryData> detection_stream = connect_map(camera_1_stream,
                    String.format("Pipeline %d: Obstacle detection", pipeline_copy), DETECTION_MSG_SIZE, false,
                    is_inter_process);

            DataStream<BinaryData> traffic_light_detection_stream = connect_map(camera_2_stream,
                    String.format("Pipeline %d: Traffic light detection", pipeline_copy), TRAFFIC_LIGHT_MSG_SIZE, false,
                    is_inter_process);

            DataStream<BinaryData> lane_detection_stream = connect_map(camera_2_stream,
                    String.format("Pipeline %d: Lane detection", pipeline_copy), LANE_DETECTION_MSG_SIZE, false,
                    is_inter_process);

            DataStream<BinaryData> tracker_stream = connect_join(camera_1_stream, detection_stream,
                    String.format("Pipeline %d: Tracker", pipeline_copy), TRACKER_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> lidar_traffic_light_fusion_stream = connect_join(lidar_stream,
                    traffic_light_detection_stream,
                    String.format("Pipeline %d: LIDAR traffic light_fusion", pipeline_copy), FUSION_MSG_SIZE,
                    is_inter_process);

            DataStream<BinaryData> tmp_fusion_stream = connect_join(tracker_stream, lidar_stream,
                    String.format("Pipeline %d: Fusion", pipeline_copy), FUSION_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> main_fusion_stream = connect_join(tmp_fusion_stream, localization_stream,
                    String.format("Pipeline %d: Fusion", pipeline_copy), FUSION_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> prediction_stream = connect_join(main_fusion_stream, localization_stream,
                    String.format("Pipeline %d: Prediction", pipeline_copy), PREDICTION_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> tmp_planning_stream = connect_join(prediction_stream, lane_detection_stream,
                    String.format("Pipeline %d: Planning", pipeline_copy), PLANNING_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> planning_stream = connect_join(tmp_planning_stream,
                    lidar_traffic_light_fusion_stream, String.format("Pipeline %d: Planning", pipeline_copy),
                    PLANNING_MSG_SIZE, is_inter_process);

            DataStream<BinaryData> control_stream = connect_join(planning_stream, localization_stream,
                    String.format("Pipeline %d: Control", pipeline_copy), CONTROL_MSG_SIZE, is_inter_process);

            String slot_sharing_group = is_inter_process ? String.format("Pipeline %d: receiver", pipeline_copy)
                    : "Source";
            DataStream<LatencySample> latency_stream = control_stream.map(new MapFunction<BinaryData, LatencySample>() {
                @Override
                public LatencySample map(BinaryData data) throws Exception {
                    double recv_time = (double) System.nanoTime() / 1e9;
                    double send_time = (double) data.send_time_ns / 1e9;
                    double serialize_start_time = (double) data.serialize_start_time_ns / 1e9;
                    double serialize_end_time = (double) data.serialize_end_time_ns / 1e9;
                    double deserialize_start_time = (double) data.deserialize_start_time_ns / 1e9;
                    double deserialize_end_time = (double) data.deserialize_end_time_ns / 1e9;
                    return new LatencySample(recv_time - send_time, send_time, recv_time, serialize_start_time,
                            serialize_end_time, deserialize_start_time, deserialize_end_time, data.data.length,
                            is_inter_process);
                }
            }).slotSharingGroup(slot_sharing_group);
            latency_streams.add(latency_stream);
        }

        DataStream<LatencySample> merged_latency_stream = latency_streams.get(0);
        for (int i = 1; i < latency_streams.size(); i++) {
            merged_latency_stream = merged_latency_stream.union(latency_streams.get(i));
        }

        String slot_sharing_group = is_inter_process ? "Sink" : "Source";
        merged_latency_stream.addSink(new CSVSink(num_warmup_msgs, csv_filename)).slotSharingGroup(slot_sharing_group);

        try {
            env.execute();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
